logic/src/commonMain/kotlin/com/wire/kalium/logic/network/NetworkStateObserver.kt

```diff
@@ -0,0 +1,54 @@
+/*
+ * Wire
+ * Copyright (C) 2023 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.network
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.withTimeoutOrNull
+import kotlin.time.Duration
+
+interface NetworkStateObserver {
+
+    fun observeNetworkState(): Flow<NetworkState>
+
+    suspend fun delayUntilConnectedWithInternetAgain(delay: Duration) {
+        // Delay for given amount but break it if reconnected again.
+        withTimeoutOrNull(delay) {
+            // Drop the current value, so it will complete only if the connection changed again to connected during that time.
+            observeNetworkState()
+                .distinctUntilChanged()
+                .drop(1)
+                .filter { it is NetworkState.ConnectedWithInternet }
+                .firstOrNull()
+        }
+        // At this point, if the block above timed out, it means that it haven't reconnected so wait until reconnected.
+        // If the block above completed before the timeout, just make sure it's still connected.
+        observeNetworkState().firstOrNull { it is NetworkState.ConnectedWithInternet }
```

Comment:
If network state is not `ConnectedWithInternet`, this will make it wait indefinitely, possibly breaking the 10min max interval.

Not sure if intended or not, if buggy or not.

Reply:
@vitorhugods that's right, that can take more than 10min, but the sync won't success anyway until we have the internet connection. It will retry when the connection state changes to `ConnectedWithInternet` so it won't try to make any requests unnecessarily. 
Would you prefer to retry sync again after the delay even if it's still not connected?

Reply:
I removed the lat "observe" - it's safer to retry every 10min even if there's no Internet than to be stuck.

Reply:
Not that I think we should retry even without internet, it's just for bug hunting purposes.

If someone ever asks themselves:
"Why isn't it respecting the 10min limit?"